File: goopylib/objects/renderable/renderable.cpp
#define GP_LOGGING_LEVEL 3

#include "goopylib/debug.h"

#include "goopylib/core/window_object.h"

#include "renderable.h"
#include "src/goopylib/objects/Renderable.h"


namespace renderable {
    PyObject *new_(PyTypeObject *type, PyObject *Py_UNUSED(args), PyObject *Py_UNUSED(kwds)) {
        GP_PY_DEBUG("gp.renderable.Renderable.__new__()");

        RenderableObject *self;
        self = (RenderableObject *) type->tp_alloc(type, 0);

        if (self != nullptr) {

        }
        return (PyObject *) self;
    }

    int init(RenderableObject *Py_UNUSED(self), PyObject *Py_UNUSED(args), PyObject *Py_UNUSED(kwds)) {
        GP_PY_INFO("gp.renderable.Renderable()");
        return 0;
    }

    PyObject *repr(RenderableObject *self) {
        GP_PY_TRACE("gp.renderable.Renderable.__repr__()");
        return PyUnicode_FromString(self->renderable->toString().c_str());
    }

    int traverse(RenderableObject *Py_UNUSED(self), visitproc Py_UNUSED(visit), void *Py_UNUSED(arg)) {
        return 0;
    }

    int clear(RenderableObject *Py_UNUSED(self)) {
        GP_PY_TRACE("gp.renderable.Renderable.clear()");
        return 0;
    }

    void dealloc(RenderableObject *self) {
        GP_PY_DEBUG("gp.renderable.Renderable.__dealloc__()");

        self->renderable.reset();

        PyObject_GC_UnTrack(self);
        clear(self);
        Py_TYPE(self)->tp_free((PyObject *) self);
    }
}

// Renderable methods
namespace renderable {
    PyObject *draw(RenderableObject *self, PyObject *arg) {
        GP_PY_DEBUG("gp.objects.renderable.Renderable.draw({0})", PyUnicode_AsUTF8(PyObject_Repr(arg)));

        #if GP_TYPE_CHECKING
        if (!isinstance(arg, WindowType)) {
            RAISE_TYPE_ERROR(nullptr, "Window", arg);
        }
        #endif

        self->renderable->draw(*((WindowObject *) arg)->window);
        RETURN_PYOBJECT((PyObject *) self);
    }

    PyObject *destroy(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        self->renderable->destroy();
        Py_RETURN_NONE;
    }

    PyObject *is_drawn(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        if (self->renderable->isDrawn()) {
            Py_RETURN_TRUE;
        }
        Py_RETURN_FALSE;
    }

    PyObject *set_anchor(RenderableObject *self, PyObject *args) {
        float x, y;
        if (!PyArg_ParseTuple(args, "ff", &x, &y)) {
            return nullptr;
        }

        self->renderable->setAnchor(x, y);
        Py_RETURN_NONE;
    }

    PyObject *reset_anchor(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        self->renderable->resetAnchor();
        Py_RETURN_NONE;
    }

    PyObject *move(RenderableObject *self, PyObject *args) {
        float x, y;
        if (!PyArg_ParseTuple(args, "ff", &x, &y)) {
            return nullptr;
        }

        self->renderable->move(x, y);
        Py_RETURN_NONE;
    }

    PyObject *rotate(RenderableObject *self, PyObject *arg) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(arg)) {
            RAISE_TYPE_ERROR(nullptr, "number", arg);
        }
        #endif

        self->renderable->rotate((float) PyFloat_AsDouble(arg));
        Py_RETURN_NONE;
    }

    PyObject *scale(RenderableObject *self, PyObject *args) {
        float x, y;

        if (PyArg_ParseTuple(args, "ff", &x, &y)) {
            self->renderable->scale(x, y);
            Py_RETURN_NONE;
        }
        PyErr_Clear();

        if (!PyArg_ParseTuple(args, "f", &x)) {
            return nullptr;
        }

        self->renderable->scale(x);
        Py_RETURN_NONE;
    }

    PyObject *set_size(RenderableObject *self, PyObject *args) {
        float width, height;
        if (!PyArg_ParseTuple(args, "ff", &width, &height)) {
            return nullptr;
        }

        self->renderable->setSize(width, height);
        Py_RETURN_NONE;
    }

    PyObject *hide(RenderableObject *self, PyObject *arg) {
        #if GP_TYPE_CHECKING
        if (!PyBool_Check(arg)) {
            RAISE_TYPE_ERROR(nullptr, "bool", arg);
        }
        #endif

        self->renderable->hide(arg == Py_True);
        Py_RETURN_NONE;
    }

    PyObject *show(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        self->renderable->show();
        Py_RETURN_NONE;
    }

    PyObject *is_hidden(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        if (self->renderable->isHidden()) {
            Py_RETURN_TRUE;
        }
        Py_RETURN_FALSE;
    }

    PyObject *is_opaque(RenderableObject *self, PyObject *Py_UNUSED(args)) {
        if (self->renderable->isOpaque()) {
            Py_RETURN_TRUE;
        }
        Py_RETURN_FALSE;
    }

    PyObject *box_contains(RenderableObject *self, PyObject *args) {
        float x, y;
        if (!PyArg_ParseTuple(args, "ff", &x, &y)) {
            return nullptr;
        }

        if (self->renderable->boxContains({x, y})) {
            Py_RETURN_TRUE;
        }
        Py_RETURN_FALSE;
    }

    PyObject *contains(RenderableObject *self, PyObject *args) {
        float x, y;
        if (!PyArg_ParseTuple(args, "ff", &x, &y)) {
            return nullptr;
        }

        if (self->renderable->contains({x, y})) {
            Py_RETURN_TRUE;
        }
        Py_RETURN_FALSE;
    }
}

// Renderable getters & setters
namespace renderable {
    // X-Position
    PyObject *get_x(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getX());
    }

    int set_x(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setX((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Y-Position
    PyObject *get_y(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getY());
    }

    int set_y(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setY((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Position
    PyObject *get_position(RenderableObject *self, void *Py_UNUSED(closure)) {
        return Py_BuildValue("ff", self->renderable->getX(), self->renderable->getY());
    }

    int set_position(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        float x, y;
        if (!PyArg_ParseTuple(value, "ff", &x, &y)) {
            return -1;
        }

        self->renderable->setPosition(x, y);
        return 0;
    }

    // Rotation
    PyObject *get_rotation(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getRotation());
    }

    int set_rotation(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setRotation((float) PyFloat_AsDouble(value));
        return 0;
    }

    // X-Scale
    PyObject *get_xscale(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getScale().xscale);
    }

    int set_xscale(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setScaleX((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Y-Scale
    PyObject *get_yscale(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getScale().yscale);
    }

    int set_yscale(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setScaleY((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Width
    PyObject *get_width(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getWidth());
    }

    int set_width(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setWidth((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Height
    PyObject *get_height(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getHeight());
    }

    int set_height(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setHeight((float) PyFloat_AsDouble(value));
        return 0;
    }

    // Z-Position
    PyObject *get_z(RenderableObject *self, void *Py_UNUSED(closure)) {
        return PyFloat_FromDouble(self->renderable->getZPosition());
    }

    int set_z(RenderableObject *self, PyObject *value, void *Py_UNUSED(closure)) {
        #if GP_TYPE_CHECKING
        if (!PyNumber_Check(value)) {
            RAISE_TYPE_ERROR(-1, "number", value);
        }
        #endif

        self->renderable->setZPosition((float) PyFloat_AsDouble(value));
        return 0;
    }
}

File: goopylib/objects/renderable/renderable.py
"""
Module defining a generic Renderable object
"""

from __future__ import annotations

from typing import Tuple, Union

from goopylib.core.window import Window


class Renderable:
    """
    This is the base class from which all goopylib objects are derived.
    The methods provided can be called by any other goopylib objects.
    """

    __slots__ = ["_renderable", "_window"]

    # pylint: disable-next=super-init-not-called
    def __init__(self) -> None:
        """
        This is the base class from which all goopylib objects are derived.
        The methods provided can be called by any other goopylib objects.

        Raises:
            NotImplementedError: cannot directly initialize a Renderable
        """
        self._renderable = None
        self._window: Union[Window, None] = None

    def __repr__(self) -> str:
        return self._renderable.__repr__()

    def draw(self, window: Window) -> Renderable:
        """
        Draws the object to a window

        Args:
            window: the Window to draw to

        Raises:
            TypeError: argument must be a Window
            RuntimeError: window has been destroyed
        """
        if isinstance(window, Window):
            self._renderable.draw(window._window)
            self._window = window
        else:
            self._renderable.draw(window)
        return self

    def destroy(self) -> None:
        """
        Destroys and undraws the object
        """
        self._renderable.destroy()

    def is_drawn(self) -> bool:
        """
        Returns:
            whether the object has been drawn
        """
        return self._renderable.is_drawn()

    def set_anchor(self, x: float, y: float) -> None:
        """
        Sets the anchor (center) of the object to the specified coordinates

        Args:
            x: in world coordinates
            y: in world coordinates

        Raises:
            TypeError: x and y must be numbers
        """
        return self._renderable.set_anchor(x, y)

    def reset_anchor(self) -> None:
        """
        Resets the anchor of the object to its center (average of its vertices)
        """
        return self._renderable.reset_anchor()

    def move(self, dx: float, dy: float) -> None:
        """
        Moves (translates) the object across the screen

        Args:
            dx: in world coordinates
            dy: in world coordinates

        Raises:
            TypeError: dx and dy must be numbers
        """
        return self._renderable.move(dx, dy)

    def rotate(self, angle: float) -> None:
        """
        Rotates the object

        Args:
            angle: in degrees

        Raises:
            TypeError: angle must be a number
        """
        return self._renderable.rotate(angle)

    def scale(self, *args: float) -> None:
        """
        Scales the object

        Args:
            *args: a float representing the scale amount. 2 floats to scale the x and y components individually.

        Raises:
            TypeError: scale factor must be numbers
        """
        return self._renderable.scale(*args)

    def set_size(self, width: float, height: float) -> None:
        """
        Sets the dimensions of the object by resizing it

        Args:
            width: in world coordinates
            height: in world coordinates

        Raises:
            TypeError: width and height must be numbers
        """
        return self._renderable.set_size(width, height)

    @property
    def window(self) -> Union[Window, None]:
        """
        The window the object is drawn to (or None)
        """
        return self._window

    @property
    def x(self) -> float:
        """
        The x-position of the object's anchor in world coordinates

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.x

    @x.setter
    def x(self, value: float) -> None:
        self._renderable.x = value

    @property
    def y(self) -> float:
        """
        The y-position of the object's anchor in world coordinates

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.y

    @y.setter
    def y(self, value: float) -> None:
        self._renderable.y = value

    @property
    def position(self) -> Tuple[float, float]:
        """
        The position of the object's anchor in world coordinates

        Returns:
            a tuple (x, y) with the position

        Raises:
            TypeError: value must be a tuple of x, y numbers
        """
        return self._renderable.position

    @position.setter
    def position(self, value: Tuple[float, float]) -> None:
        self._renderable.position = value

    @property
    def rotation(self) -> float:
        """
        The rotation of the object in degrees

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.rotation

    @rotation.setter
    def rotation(self, value: float) -> None:
        self._renderable.rotation = value

    @property
    def xscale(self) -> float:
        """
        The x-scale of the object in %

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.xscale

    @xscale.setter
    def xscale(self, value: float) -> None:
        self._renderable.xscale = value

    @property
    def yscale(self) -> float:
        """
        The y-scale of the object in %

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.yscale

    @yscale.setter
    def yscale(self, value: float) -> None:
        self._renderable.yscale = value

    @property
    def width(self) -> float:
        """
        The width of the object in world coordinates

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.width

    @width.setter
    def width(self, value: float) -> None:
        self._renderable.width = value

    @property
    def height(self) -> float:
        """
        The height of the object in world coordinates

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.height

    @height.setter
    def height(self, value: float) -> None:
        self._renderable.height = value

    @property
    def z(self) -> float:
        """
        The z-position of the object

        Raises:
            TypeError: value must be a number
        """
        return self._renderable.z

    @z.setter
    def z(self, value: float) -> None:
        self._renderable.z = value

    def hide(self, hide=True) -> None:
        """
        Hides the object from the window

        Args:
            hide: whether to hide or show

        Raises:
            TypeError: hide must be a boolean
        """
        return self._renderable.hide(hide)

    def show(self) -> None:
        """
        Unhides the object if it was hidden
        """
        return self._renderable.show()

    def is_hidden(self) -> bool:
        """
        Returns:
            whether the object is hidden
        """
        return self._renderable.is_hidden()

    def is_opaque(self) -> bool:
        """
        Returns:
            whether the object is completely opaque
        """
        return self._renderable.is_opaque()

    def box_contains(self, x: float, y: float) -> bool:
        """
        Checks if the object's rectangular bounding box contains a point

        Args:
            x: in world coordinates
            y: in world coordinates

        Raises:
            TypeError: x and y must be numbers
        """
        return self._renderable.box_contains(x, y)

    def contains(self, x: float, y: float) -> bool:
        """
        Checks if the object's bounding box contains a point

        Args:
            x: in world coordinates
            y: in world coordinates

        Raises:
            TypeError: x and y must be numbers
        """
        return self._renderable.contains(x, y)
